File: benchmarks/tasks/contention/contention.py
"""Threads contending for shared mutable state: eight of them moving money
between sixteen accounts, every transfer reading two accounts and writing two
as one indivisible step. One lock over the whole set.

CPython's own lock means these threads take turns rather than run at once, so
this measures what the locking costs and not what parallelism buys."""

import threading

ACCOUNTS = 16
WORKERS = 8
MOVES = 20000

bank = [1000] * ACCOUNTS
lock = threading.Lock()


def transfers(w):
    s = w + 1
    for _ in range(MOVES):
        s = s * 48271 % 2147483647
        a = s % ACCOUNTS
        b = s // ACCOUNTS % ACCOUNTS
        amount = 1 + s % 10
        if a == b:
            continue
        with lock:
            bank[a] -= amount
            bank[b] += amount


def main():
    threads = [threading.Thread(target=transfers, args=(w,)) for w in range(WORKERS)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    print(",".join(str(b) for b in bank))


if __name__ == "__main__":
    main()

File: benchmarks/tasks/contention/Contention.java
// Threads contending for shared mutable state: eight of them moving money
// between sixteen accounts, every transfer reading two accounts and writing
// two as one indivisible step. One monitor over the whole set.

import java.util.StringJoiner;

public class Contention {
    static final int ACCOUNTS = 16;
    static final int WORKERS = 8;
    static final int MOVES = 20_000;

    static final long[] bank = new long[ACCOUNTS];
    static final Object lock = new Object();

    public static void main(String[] args) throws InterruptedException {
        java.util.Arrays.fill(bank, 1000);
        Thread[] threads = new Thread[WORKERS];
        for (int w = 0; w < WORKERS; w++) {
            final long seed = w + 1;
            threads[w] = new Thread(() -> {
                long s = seed;
                for (int i = 0; i < MOVES; i++) {
                    s = s * 48271 % 2147483647;
                    int a = (int) (s % ACCOUNTS);
                    int b = (int) (s / ACCOUNTS % ACCOUNTS);
                    long amount = 1 + s % 10;
                    if (a == b) continue;
                    synchronized (lock) {
                        bank[a] -= amount;
                        bank[b] += amount;
                    }
                }
            });
            threads[w].start();
        }
        for (Thread t : threads) t.join();
        StringJoiner out = new StringJoiner(",");
        for (long b : bank) out.add(Long.toString(b));
        System.out.println(out);
    }
}
